common/common-ui/src/main/java/com/duckduckgo/common/ui/experiments/visual/store/VisualDesignExperimentDataStoreImpl.kt

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2025 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.common.ui.experiments.visual.store
+
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.booleanPreferencesKey
+import androidx.datastore.preferences.core.edit
+import com.duckduckgo.app.di.AppCoroutineScope
+import com.duckduckgo.common.ui.experiments.visual.ExperimentalUIThemingFeature
+import com.duckduckgo.common.ui.experiments.visual.store.VisualDesignExperimentDataStore.FeatureState
+import com.duckduckgo.di.scopes.AppScope
+import com.duckduckgo.feature.toggles.api.Toggle
+import com.duckduckgo.privacy.config.api.PrivacyConfigCallbackPlugin
+import com.squareup.anvil.annotations.ContributesBinding
+import com.squareup.anvil.annotations.ContributesMultibinding
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+
+@ContributesBinding(
+    scope = AppScope::class,
+    boundType = VisualDesignExperimentDataStore::class,
+)
+@ContributesMultibinding(scope = AppScope::class, boundType = PrivacyConfigCallbackPlugin::class)
+class VisualDesignExperimentDataStoreImpl @Inject constructor(
+    @VisualDesignExperiment private val store: DataStore<Preferences>,
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+    experimentalUIThemingFeature: ExperimentalUIThemingFeature,
+) : VisualDesignExperimentDataStore, PrivacyConfigCallbackPlugin {
+    private companion object {
+
+        private const val KEY_EXPERIMENT_ENABLED = "KEY_EXPERIMENT_ENABLED"
+        private const val DEFAULT_EXPERIMENT_USER_PREFERENCE = true
+
+        private const val KEY_NAVIGATION_BAR_ENABLED = "KEY_NAVIGATION_BAR_ENABLED"
+        private const val DEFAULT_NAVIGATION_BAR_USER_PREFERENCE = true
```

Comment:
This governs whether the features are enabled by default.

In our case, we're shipping to `internal` with a `@InternalAlwaysEnabled`, so the feature flag will always appear turned on but its only this preference which will actually apply it. The preference allows user to turn the feature off and it allows us to turn the feature completely off remotely or with an app update.